test/queues.c

```diff
@@ -166,8 +166,116 @@ TEST test_queue_teardown(void)
         PASS();
 }
 
+TEST test_datachange_beginning_empty(void)
+{
+        queues_init();
+
+        ASSERTm("There are no notifications at all, the timeout has to be less than 0.",
+                queues_get_next_datachange(time_monotonic_now()) < 0);
+
+        queues_teardown();
+        PASS();
+}
+
+TEST test_datachange_endless(void)
+{
+        queues_init();
+
+        settings.show_age_threshold = -1;
+
+        struct notification *n = test_notification("n", 0);
+
+        queues_notification_insert(n);
+        queues_update(STATUS_NORMAL);
+
+        ASSERTm("Age threshold is deactivated and the notification is infinite, there is no wakeup necessary.",
+                queues_get_next_datachange(time_monotonic_now()) < 0);
+
+        queues_teardown();
+        PASS();
+}
+
+TEST test_datachange_endless_agethreshold(void)
+{
+        settings.show_age_threshold = S2US(5);
+
+        queues_init();
+
+        struct notification *n = test_notification("n", 0);
+
+        queues_notification_insert(n);
+        queues_update(STATUS_NORMAL);
+
+        ASSERT_IN_RANGEm("Age threshold is activated and the next wakeup should be less than a second away",
+                S2US(1), queues_get_next_datachange(time_monotonic_now() + S2US(4)), S2US(1));
+
+        ASSERT_IN_RANGEm("Age threshold is activated and the next wakeup should be less than the age threshold",
+                settings.show_age_threshold, queues_get_next_datachange(time_monotonic_now()), settings.show_age_threshold);
+
+        settings.show_age_threshold = S2US(0);
+        ASSERT_IN_RANGEm("Age threshold is activated and the next wakeup should be less than a second away",
+                S2US(1), queues_get_next_datachange(time_monotonic_now()), S2US(1));
+
+        queues_teardown();
+        PASS();
+}
+
+TEST test_datachange_queues(void)
+{
+        queues_init();
+
+        struct notification *n = test_notification("n", 10);
+
+        queues_notification_insert(n);
+        ASSERTm("The inserted notification is inside the waiting queue, so it should get ignored.",
+               queues_get_next_datachange(time_monotonic_now()) < S2US(0));
+
+        queues_update(STATUS_NORMAL);
+        ASSERT_IN_RANGEm("The notification has to get closed in less than its timeout",
+               S2US(10), queues_get_next_datachange(time_monotonic_now()), S2US(10));
```

Comment:
`ASSERT_IN_RANGE` doesn't make sense here, for one it assumes the value is double which makes the error message confusing by using scientific notation and secondly if `S2US(15)` is returned it still passes as it checks for `+/- S2US(10)`.

Reply:
Oh, well it **does** make sense. Most importantly it has to be less than the actual timeout. But also it should be non-negative (this would be the representation for not sleeping).